src/bsoncxx/test/bson_serialization.cpp

```diff
@@ -0,0 +1,322 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include <bsoncxx/builder/basic/array.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/document/value.hpp>
+#include <bsoncxx/string/to_string.hpp>
+#include <bsoncxx/test_util/catch.hh>
+#include <bsoncxx/types/bson_value/view.hpp>
+
+using namespace bsoncxx;
+
+using builder::basic::kvp;
+using builder::basic::make_array;
+using builder::basic::make_document;
+
+namespace {
+
+namespace test {
+
+struct Person {
+    std::string first_name;
+    std::string last_name;
+    int age;
+};
+
+// to_bson serializer function that can take a document::value&
+void to_bson(const Person& person, bsoncxx::document::value& bson_object) {
+    bson_object = make_document(kvp("first_name", person.first_name),
+                                kvp("last_name", person.last_name),
+                                kvp("age", person.age));
+}
+
+void from_bson(Person& person, const bsoncxx::document::view& bson_object) {
+    bson_object["first_name"].get_value().to_field(person.first_name);
+    bson_object["last_name"].get_value().to_field(person.last_name);
+    bson_object["age"].get_value().to_field(person.age);
+}
+
+class Car {
+   public:
+    Car() = default;
+    Car(std::string manufacturer, std::string model)
+        : _manufacturer(std::move(manufacturer)), _model(std::move(model)){};
+
+    std::string get_manufacturer() const {
+        return _manufacturer;
+    }
+
+    std::string get_model() const {
+        return _model;
+    }
+
+    friend void to_bson(const Car& car, bsoncxx::document::value& bson_object) {
+        auto builder = bsoncxx::builder::stream::document{};
+        bson_object = builder << "manufacturer" << car._manufacturer << "model" << car._model
+                              << bsoncxx::builder::stream::finalize;
+    }
+
+    friend void from_bson(Car& car, const bsoncxx::document::view& bson_object) {
+        bson_object["manufacturer"].get_value().to_field(car._manufacturer);
+        bson_object["model"].get_value().to_field(car._model);
+    }
+
+   private:
+    std::string _manufacturer;
+    std::string _model;
+};
+
+// Class that uses most BSON Types
+class Types_Test_Object {
```

Comment:
Echoing the comment above, I think the only class we need here is `Person` and its corresponding tests.